src/runtime/rpc/group_host_port.h

```diff
@@ -0,0 +1,244 @@
+    set_leader(host_port(g_addr->leader()));
+}
+
+inline rpc_group_host_port &rpc_group_host_port::operator=(const rpc_group_host_port &other)
+{
+    if (this == &other) {
+        return *this;
+    }
+    _name = other._name;
+    _leader_index = other._leader_index;
+    _update_leader_automatically = other._update_leader_automatically;
+    _members = other._members;
+    return *this;
+}
+
+inline bool rpc_group_host_port::add(host_port hp)
+{
+    CHECK_EQ_MSG(hp.type(), HOST_TYPE_IPV4, "rpc group host_port member must be ipv4");
+
+    alw_t l(_lock);
+    if (_members.end() == std::find(_members.begin(), _members.end(), hp)) {
+        _members.push_back(hp);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+inline void rpc_group_host_port::leader_forward()
+{
+    alw_t l(_lock);
+    if (_members.empty()) {
+        return;
+    }
+    _leader_index = (_leader_index + 1) % _members.size();
+}
+
+inline void rpc_group_host_port::set_leader(host_port hp)
+{
+    CHECK_EQ_MSG(hp.type(), HOST_TYPE_IPV4, "rpc group host_port member must be ipv4");
+    alw_t l(_lock);
+    if (hp.is_invalid()) {
+        _leader_index = kInvalidIndex;
+        return;
+    }
+    for (int i = 0; i < _members.size(); i++) {
+        if (_members[i] == hp) {
+            _leader_index = i;
+            return;
+        }
+    }
+
+    _members.push_back(hp);
+    _leader_index = static_cast<int>(_members.size() - 1);
+}
+
+inline host_port rpc_group_host_port::possible_leader()
+{
+    alw_t l(_lock);
+    if (_members.empty()) {
+        return host_port::s_invalid_host_port;
+    }
+    if (_leader_index == kInvalidIndex) {
+        _leader_index = rand::next_u32(0, static_cast<uint32_t>(_members.size() - 1));
+    }
+    return _members[_leader_index];
+}
+
+inline bool rpc_group_host_port::remove(host_port hp)
+{
+    alw_t l(_lock);
+    auto it = std::find(_members.begin(), _members.end(), hp);
+    if (it == _members.end()) {
+        return false;
+    }
+
+    if (kInvalidIndex != _leader_index && hp == _members[_leader_index]) {
+        _leader_index = kInvalidIndex;
+    }
+
+    _members.erase(it);
+
+    return true;
+}
+
+inline bool rpc_group_host_port::contains(host_port hp) const
+{
+    alr_t l(_lock);
+    return _members.end() != std::find(_members.begin(), _members.end(), hp);
+}
+
+inline int rpc_group_host_port::count() const
+{
+    alr_t l(_lock);
+    return _members.size();
+}
+
+inline host_port rpc_group_host_port::next(host_port current) const
+{
+    alr_t l(_lock);
+    if (_members.empty()) {
+        return host_port::s_invalid_host_port;
+    }
+
+    if (current.is_invalid()) {
+        return _members[rand::next_u32(0, static_cast<uint32_t>(_members.size() - 1))];
+    }
+
+    auto it = std::find(_members.begin(), _members.end(), current);
+    if (it == _members.end()) {
+        return _members[rand::next_u32(0, static_cast<uint32_t>(_members.size() - 1))];
```

Comment:
`rand::next_u32(0, static_cast<uint32_t>(_members.size() - 1))` apears 4 times, let's define it as a function like `uint32_t random_index_unlocked()`?